server/service/src/document/document_service.rs

```diff
@@ -134,6 +216,100 @@ fn validate_parents(
     Ok(None)
 }
 
+fn validate_document_delete(
+    connection: &StorageConnection,
+    id: &str,
+) -> Result<Document, DocumentDeleteError> {
+    let doc = match DocumentRepository::new(connection).find_one_by_id(id)? {
+        Some(doc) => {
+            if doc.status == DocumentStatus::Active {
+                doc
+            } else {
+                return Err(DocumentDeleteError::CannotDeleteDeletedDocument);
+            }
+        }
+        None => {
+            return Err(DocumentDeleteError::DocumentNotFound);
+        }
+    };
+    Ok(doc)
+}
+
+fn validate_document_undelete(
+    connection: &StorageConnection,
+    id: &str,
+) -> Result<Document, DocumentUndeleteError> {
+    let doc = match DocumentRepository::new(connection).find_one_by_id(id)? {
+        Some(doc) => {
+            if doc.status == DocumentStatus::Deleted {
+                doc
+            } else {
+                return Err(DocumentUndeleteError::CannotUndeleteActiveDocument);
+            }
+        }
+        None => {
+            return Err(DocumentUndeleteError::DocumentNotFound);
+        }
+    };
+    Ok(doc)
+}
+
+fn generate_deleted_document(
+    input: DocumentDelete,
+    current_document: Document,
+    user_id: &str,
+) -> Result<Document, DocumentDeleteError> {
+    let doc = RawDocument {
+        name: current_document.name,
+        parents: vec![current_document.id.clone()],
+        author: user_id.to_string(),
+        timestamp: Utc::now(),
+        r#type: current_document.r#type,
+        data: serde_json::Value::Null,
+        schema_id: current_document.schema_id,
+        status: DocumentStatus::Deleted,
+        comment: input.comment,
+    }
+    .finalise()
+    .map_err(|err| DocumentDeleteError::InternalError(err))?;
+
+    Ok(doc)
+}
+
+fn generate_undeleted_document(
+    connection: &StorageConnection,
+    current_document: Document,
+    user_id: &str,
+) -> Result<Document, DocumentUndeleteError> {
+    let parent = match current_document.parent_ids.last() {
+        Some(parent) => parent,
+        None => "",
+    };
+
+    let deleted_document_parent =
+        match DocumentRepository::new(connection).find_one_by_id(&parent.clone()) {
+            Ok(Some(doc)) => doc,
+            Ok(None) => return Err(DocumentUndeleteError::ParentDoesNotExist),
+            Err(err) => return Err(DocumentUndeleteError::DatabaseError(err)),
+        };
```

Comment:
I think elsewhere we would put this into the validate method, i.e. verify that the parent exists and then return the current_document and the original_document from validate()...

Reply:
I don't think it needs to return current document since validation is already used on the deleted document and only the id from the deleted document is needed for generating undeleted document 
[f2750ed](https://github.com/openmsupply/open-msupply/pull/521/commits/f2750edacd6b02fecba686114219adced5a86239)